modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java

```diff
@@ -0,0 +1,889 @@
+        public final int grpId;
+
+        /** */
+        public final int partId;
+
+        /** */
+        public final DataRegion cacheDataRegion;
+
+        /** */
+        public final PageMemoryEx cachePageMemory;
+
+        /** */
+        public final PageMemoryEx partPageMemory;
+
+        /** */
+        public final PageMemoryEx mappingPageMemory;
+
+        /** */
+        public final CacheGroupContext oldGrpCtx;
+
+        /** */
+        public final CacheGroupContext newGrpCtx;
+
+        /** */
+        public final CacheDataStore oldCacheDataStore;
+
+        /** */
+        private GridCacheDataStore newCacheDataStore;
+
+        /** */
+        public final FilePageStoreFactory pageStoreFactory;
+
+        /** */
+        public final AtomicLong partPagesAllocated = new AtomicLong();
+
+        /** */
+        public final AtomicLong mappingPagesAllocated = new AtomicLong();
+
+        /** */
+        private LinkMap linkMap;
+
+        /** */
+        public PartitionContext(
+            File workDir,
+            int grpId,
+            int partId,
+            DataRegion partDataRegion,
+            DataRegion mappingDataRegion,
+            CacheGroupContext oldGrpCtx,
+            CacheGroupContext newGrpCtx,
+            CacheDataStore oldCacheDataStore,
+            FilePageStoreFactory pageStoreFactory
+        ) {
+            this.workDir = workDir;
+            this.grpId = grpId;
+            this.partId = partId;
+            cacheDataRegion = oldGrpCtx.dataRegion();
+
+            cachePageMemory = (PageMemoryEx)cacheDataRegion.pageMemory();
+            partPageMemory = (PageMemoryEx)partDataRegion.pageMemory();
+            mappingPageMemory = (PageMemoryEx)mappingDataRegion.pageMemory();
+
+            this.oldGrpCtx = oldGrpCtx;
+            this.newGrpCtx = newGrpCtx;
+            this.oldCacheDataStore = oldCacheDataStore;
+            this.pageStoreFactory = pageStoreFactory;
+        }
+
+        /** */
+        public PageStore createPartPageStore() throws IgniteCheckedException {
+            PageStore partPageStore;
+
+            defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadLock();
+            try {
+                partPageStore = pageStoreFactory.createPageStore(
+                    FLAG_DATA,
+                    () -> defragmentedPartTmpFile(workDir, partId).toPath(),
+                    partPagesAllocated::addAndGet
+                );
+            }
+            finally {
+                defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadUnlock();
+            }
+
+            partPageStore.sync();
+
+            DefragmentationPageReadWriteManager pageMgr = (DefragmentationPageReadWriteManager)partPageMemory.pageManager();
+
+            pageMgr.pageStoreMap().addPageStore(grpId, partId, partPageStore);
+
+            return partPageStore;
+        }
+
+        /** */
+        public PageStore createMappingPageStore() throws IgniteCheckedException {
```

Comment:
`createMappingPageStore`, `createPartPageStore` are almost the same. Is it good practice having such code duplication?

Reply:
I'll join them